Tasks/FtpUploadV2/ftpuploadtask.ts

```diff
@@ -23,142 +20,322 @@ export class FtpOptions {
+    if (win) {
+        matchOptions["nocase"] = true;
+    }
+
+    tl.debug("Candidates found for match: " + allFiles.length);
+    for (let i = 0; i < allFiles.length; i++) {
+        tl.debug("file: " + allFiles[i]);
+    }
+
+    // use a set to avoid duplicates
+    let matchingFilesSet: Set<string> = new Set();
+
+    for (let i = 0; i < ftpOptions.filePatterns.length; i++) {
+        let normalizedPattern: string = path.join(
+            ftpOptions.rootFolder,
+            path.normalize(ftpOptions.filePatterns[i])
+        );
+
+        tl.debug("searching for files, pattern: " + normalizedPattern);
+
+        let matched = tl.match(allFiles, normalizedPattern, null, matchOptions);
+        tl.debug("Found total matches: " + matched.length);
+        // ensure each result is only added once
+        for (let j = 0; j < matched.length; j++) {
+            let match = path.normalize(matched[j]);
+            let stats = tl.stats(match);
+            if (!ftpOptions.preservePaths && stats.isDirectory()) {
+                // if not preserving paths, skip all directories
+            } else if (matchingFilesSet.add(match)) {
+                tl.debug(
+                    "adding " + (stats.isFile() ? "file:   " : "folder: ") + match
+                );
+                if (stats.isFile() && ftpOptions.preservePaths) {
+                    // if preservePaths, make sure the parent directory is also included
+                    let parent = path.normalize(path.dirname(match));
+                    if (matchingFilesSet.add(parent)) {
+                        tl.debug("adding folder: " + parent);
+                    }
+                }
+            }
         }
-    });
+    }
+    return Array.from(matchingFilesSet).sort();
+}
 
-    ftpClient.on('close', (hadErr: boolean) => {
-        console.log(tl.loc('Disconnected'));
-        tl.debug('ftp client close, hadErr:' + hadErr);
-    });
+async function run() {
+    tl.setResourcePath(path.join(__dirname, "task.json"));
+
+    const tries = 3;
+    const ftpOptions: FtpOptions = getFtpOptions();
 
-    ftpClient.on('end', () => {
-        tl.debug('ftp client end');
-    })
+    if (!ftpOptions.serverEndpointUrl.protocol) {
+        tl.setResult(tl.TaskResult.Failed, tl.loc("FTPNoProtocolSpecified"));
+    }
+    if (!ftpOptions.serverEndpointUrl.hostname) {
+        tl.setResult(tl.TaskResult.Failed, tl.loc("FTPNoHostSpecified"));
+    }
 
-    ftpClient.on('error', (err) => {
-        tl.debug('ftp client error, err: ' + err);
-        if (!uploadSuccessful) {
-            // once all files are successfully uploaded, a subsequent error should not fail the task
-            failTask(err);
+    const files: string[] = findFiles(ftpOptions);
+    let ftpClient = await getFtpClient(ftpOptions);
+    const tracker = new ProgressTracking(ftpOptions, files.length + 1); // add one extra for the root directory
+
+    let retryWithNewClient = async (task: () => {}, retry: number) => {
```

Comment:
If I were to see this code for the first time, I would wonder why we did this.  Could you please add some comments here as an explanation?

Reply:
I like the "retry" function, except that it has special knowledge of the FTP client. A more modular approach could take another function, `onFailure`, that in this case looks like this:

```ts
() => {
    ftpClient.close();
    await sleep(1000);
    ftpClient = await getFtpClient(ftpOptions);
}
```

Reply:
Also confused about the "new client" here. Don't we make the call with the same type of client every time?

Reply:
Right, technically you don't need a new client, you just need to close the connection and reconnect I assume.  I just went with a new client to make sure we start clean.